src/cgutils.cpp

```diff
@@ -941,13 +941,46 @@ static void emit_type_error(const jl_cgval_t &x, Value *type, const std::string
 #endif
 }
 
+// Type{...} and Union{} are never constructable
+// so there is no value `x` s.t. isa(x, type) is true
+// because `type <: typeof(x)`
+// so `type <: x.typ` does not imply `isa(x, type)` for these cases
+static bool verify_type_is_constructable(jl_value_t *type)
+{
+    if (jl_is_tuple_type(type)) {
+        jl_svec_t *tt = ((jl_tupletype_t*)type)->types;
+        size_t i, l = jl_svec_len(tt);
+        for (i = 0; i < l; i++) {
+            if (!verify_type_is_constructable(jl_svecref(tt, i)))
+                return false;
+        }
+        return true;
+    }
+    else if (type == jl_bottom_type) {
+        return false;
+    }
+    else if (jl_is_type_type(type)) {
+        return false;
+    }
+    else if (jl_is_uniontype(type)) {
+        jl_uniontype_t *u = (jl_uniontype_t*)type;
+        return verify_type_is_constructable(u->a) && verify_type_is_constructable(u->b);
```

Comment:
I think this would have to be `||` instead of `&&`. I also think this whole function should not be necessary; we can leave these cases to constant folding and `jl_isa`.